wallet: Allow negative effective value inputs when subtracting fee from outputs

achow101 opened the pull request:
Often the goal with subtracting fee from outputs is to sweep an entire wallet's balance. As such, it is reasonable to allow negative effective value UTXOs as inputs as otherwise the target value (the balance) cannot be reached. However this is not always desirable as there may be solutions with only positive EV UTXOs, so we only try with negative EV UTXOs if we are unable to find any solutions with the normal groupings.

Fixes #23026

achow101 committed:
wallet: Set change fee only if not subtracting fee from outputs

achow101 committed:
tests: Test for spending neg ev UTXOs when subtracting fee from outputs

When subtracting fee from outputs, it should be okay to spend negative
ev UTXOs because the recipient is paying for it.

fanquake commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -1143,6 +1144,21 @@ def test_feerate_rounding(self):
+        created = w.createrawtransaction([], [{self.nodes[0].getnewaddress(address_type="bech32"): 1.00001000}])
+
+        # With 15 sat/vb, the low value input is negative ev. This should still fund.
+        funded = w.fundrawtransaction(created, {"subtractFeeFromOutputs":[0], "fee_rate": 15})
```
`test/functional/rpc_fundrawtransaction.py:1161:9: F841 local variable 'funded' is assigned to but never used`

achow101 replied:
Fixed.

meshcollider reviewed (commented):
Code review ACK c351e53f42c85c7ed95bde842f07554c9798dbcb

Checked that the new test fails on current master but passes with this PR.

S3RK commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -1143,6 +1144,21 @@ def test_feerate_rounding(self):
         rawtx = w.createrawtransaction(inputs=[], outputs=[{self.nodes[0].getnewaddress(address_type="bech32"): 1 - 0.00000202}])
         assert_raises_rpc_error(-4, "Insufficient funds", w.fundrawtransaction, rawtx, {"fee_rate": 1.85})
 
+    def test_sffo_with_negative_ev(self):
+        self.log.info("Testing that subtract fee from outputs with negative effective value outputs still works")
```
```suggestion
    def test_wallet_sweaping(self):
        self.log.info("Testing that wallet sweeping works even with negative effective value.")
```

instagibbs replied:
the functionality change isn't for sweeping per-se, maybe just add "often used for sweeping" in the comment

rajarshimaitra replied:
I am finding this line a little a confusing, maybe something like "Testing that sweeping transaction with "subtract fee from outputs" with negative effective value utxos works".

achow101 replied:
This functionality is for more than just sweeping. I've changed the text to `Test that enable subtractFeeFromOutputs will include negative effective value UTXOs`.

S3RK commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -1143,6 +1144,21 @@ def test_feerate_rounding(self):
+        created = w.createrawtransaction([], [{self.nodes[0].getnewaddress(address_type="bech32"): 1.00001000}])
+
+        # With 15 sat/vb, the low value input is negative ev. This should still fund.
+        w.fundrawtransaction(created, {"subtractFeeFromOutputs":[0], "fee_rate": 15})
```
I'd add another assertion that it fails without `subtractFeeFromOutputs`

instagibbs replied:
can we make 15 more programmatic with respect to the 1000 sat output?

achow101 replied:
Added an `assert_raises_rpc_error`.

Also made the fee rate calculation programmatic

S3RK commented:
Concept ACK. 

Did a light code review and I think it's better to drop the first commit as it's not required to fix the linked bug. If you'd like to keep it, I'd suggest to make inclusion/exclusion of the change fee explicit when stetting targets for coinselection algorithms, rather than implicit `m_change_fee = 0` (it's not zero, it's just payed by recipient!).

S3RK reviewed (commented):
some nits below

gruve-p commented:
ACK https://github.com/bitcoin/bitcoin/pull/23534/commits/c351e53f42c85c7ed95bde842f07554c9798dbcb

fanquake commented:
cc @yurayakimenko

instagibbs commented on src/wallet/spend.cpp:
```diff
@@ -384,7 +384,7 @@ bool AttemptSelection(const CWallet& wallet, const CAmount& nTargetValue, const
 
     // Note that unlike KnapsackSolver, we do not include the fee for creating a change output as BnB will not create a change output.
-    std::vector<OutputGroup> positive_groups = GroupOutputs(wallet, coins, coin_selection_params, eligibility_filter, true /* positive_only */);
+    std::vector<OutputGroup> positive_groups = GroupOutputs(wallet, coins, coin_selection_params, eligibility_filter, !coin_selection_params.m_subtract_fee_outputs/* positive_only */);
```
positive_groups is a misnomer, perhaps rename it `non_knapsack_groups`?

achow101 replied:
Named this to just `groups` and the groups that `KnapsackSolver` uses to be `knapsack_groups`. It's better for the special case to be named after the special case rather than the normal case to be named as the negation of the special case.

instagibbs commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -1143,6 +1144,21 @@ def test_feerate_rounding(self):
+        self.log.info("Testing that subtract fee from outputs with negative effective value outputs still works")
+
+        self.nodes[1].createwallet("negevsffotest")
+        w = self.nodes[1].get_wallet_rpc("negevsffotest")
```
assert `w` has no balance/utxos, may seem duh but it helps readability of the two wallet dance

achow101 replied:
Done

instagibbs commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -1143,6 +1144,21 @@ def test_feerate_rounding(self):
+        self.nodes[1].createwallet("negevsffotest")
+        w = self.nodes[1].get_wallet_rpc("negevsffotest")
+
+        self.nodes[0].sendtoaddress(w.getnewaddress(address_type="bech32"), 0.00001000)
```
just make this 1 sat?

instagibbs commented:
concept ACK, if someone wants to do "send max value possible" we could support this as another option and choose positive only groups again.

rajarshimaitra reviewed (commented):
tACK https://github.com/bitcoin/bitcoin/pull/23534/commits/c351e53f42c85c7ed95bde842f07554c9798dbcb

Verified that the test fails in master.

Though I am thinking, the problem of negative ev utxos are only relevant for sweeping transactions with `fundrawtransaction` rpc (there are other ways to create sweeps). Does this warrant removing negative ev for all cases with `subtract_fee_outputs`? Maybe we can have a more sweeping focused approach to trigger BnB?

luke-jr reviewed (changes requested):
But this solution means you will end up with less bitcoins on the receiving end, doesn't it? Users *probably* would prefer the higher amount received.

I suppose the best UX would be to either ignore the net-negative UTXOs and simply send less than the requested amount. But that is apt to unprovably "burn" the dust unless we can somehow publish the dust private keys for others to sweep (but that seems dangerous and potentially ties to some centralised publisher).

Second-best might be to effectively reduce the fee rate by simply refusing to pay more for these UTXOs than they're worth. But care is needed to ensure the user is aware of this and transactions don't get stuck.

Also, any of this should only occur when trying to send the entire balance I think?

achow101 commented:
> But this solution means you will end up with less bitcoins on the receiving end, doesn't it?

Yes.

> Users _probably_ would prefer the higher amount received.

I'm not sure. Subtract fee from outputs has always been "X amount leaves the wallet", so I think having less than that leave would be surprising? 

> I suppose the best UX would be to either ignore the net-negative UTXOs and simply send less than the requested amount. But that is apt to unprovably "burn" the dust unless we can somehow publish the dust private keys for others to sweep (but that seems dangerous and potentially ties to some centralised publisher).

Negative EV UTXOs can become positive if the feerate were to drop (assuming it was high).

> Second-best might be to effectively reduce the fee rate by simply refusing to pay more for these UTXOs than they're worth. But care is needed to ensure the user is aware of this and transactions don't get stuck.

I think that might be dangerous, and surprising to users.

> Also, any of this should only occur when trying to send the entire balance I think?

I think it is reasonable for this to occur for all subtract fee from outputs sends. Maybe it should try positive only first, and if that fails, allow negative?

instagibbs commented:
ACK https://github.com/bitcoin/bitcoin/pull/23534/commits/f44c3ce31921d15dfae5375b74910b350003fbb4

Sjors commented:
Nit: typo in the PR title "negtive" (may make it harder to find later)

I typically use the "subtract fee" feature in combination with manual coin selection (for which this PR has no effect). That's different from the use case of emptying out a wallet. At minimum the RPC docs should warn that this will spend negative net value coins. The GUI should also warn, preferably only when this is actually the case. It can get quite expensive if a wallet has lots of dust.

achow101 committed:
wallet: Try selecting with negative ev coins when subtractFeeFromOutputs

Sometimes subtractFeeFromOutputs requires spending negative ev UTXOs
(e.g. when sweeping an entire wallet). If we are unable to find any
solutions normally, then try again but this time allowing negative ev
UTXOs.

achow101 commented:
I've changed this to try negative EVs if there are no solutions with the normal groupings.

meshcollider reviewed (commented):
Code review ACK a26a64c63dd8d55d0ac972f5420f5dd27e7f53d8

Haven't re-verified test

instagibbs commented on src/wallet/spend.cpp:
```diff
@@ -403,6 +403,20 @@ std::optional<SelectionResult> AttemptSelection(const CWallet& wallet, const CAm
 
+    // If we didn't find any solutions, and are subtracting the fee from the outputs, we may need to include negative effective value coins.
+    // So try BnB and SRD again with all_groups.
+    // It is safe to do this because a solution including negative EVs will have a greater waste.
```
What about just throwing these two selections into the mix unconditionally? Would that be correct? Efficiency aside.

instagibbs reviewed (approved):
good change to simplify the behavior difference

reACK https://github.com/bitcoin/bitcoin/pull/23534/commits/a26a64c63dd8d55d0ac972f5420f5dd27e7f53d8

Xekyo commented:
I already mentioned this in person, but I find that "subtract fee" leads to unintuitive behavior, especially in the context of sweeping (and makes all tests a lot more complicated).

It would make more sense to me that the wallet _never_ spends UTXOs with negative effective value unless explicitly tasked to do so, since it just costs users (either sender or receiver) more money for no good reason. The main use-case where I see "subtract fee from recipient" making sense, is in the context of Coin Control, when the sender picks the inputs explicitly and doesn't want to calculate the appropriate fee manually.

Especially in the case of sweeping, I agree with Luke, that the user should prefer getting more money over using all UTXOs. If they sweep at `minFeeRate`, they're getting the optimal outcome anyway. We might want to introduce a dedicated sweep function that sends the maximum by default, and optionally can be toggled to spend all UTXOs.

ryanofsky commented:
> I already mentioned this in person, but I find that "subtract fee" leads to unintuitive behavior, especially in the context of sweeping (and makes all tests a lot more complicated).

Could you maybe illustrate what the unintuitive behavior is with an example? The rest of your comment about "user should get more money by default" obviously sounds good if it doesn't have other tradeoffs.

achow101 commented:
> The main use-case where I see "subtract fee from recipient" making sense, is in the context of Coin Control, when the sender picks the inputs explicitly and doesn't want to calculate the appropriate fee manually.

And that scenario could be considered sweeping as well. Sweeping can be generalized to "spend UTXOs XYZ fully with all value going to one output" where XYZ could be all UTXOs in the wallet, or a preset input list.

> Especially in the case of sweeping, I agree with Luke, that the user should prefer getting more money over using all UTXOs. If they sweep at `minFeeRate`, they're getting the optimal outcome anyway. We might want to introduce a dedicated sweep function that sends the maximum by default, and optionally can be toggled to spend all UTXOs.

I think it is unintuitive and not very good UX to not spend uneconomical UTXOs in a sweep. I think that users would expect a sweep function to spend all UTXOs regardless, and it would be surprising if sweeping a wallet left behind any balance, even uneconomical.

> Could you maybe illustrate what the unintuitive behavior is with an example?

I'm not sure about unintuitive, but subtractFeeFromOutputs definitely causes a lot of headaches in coin selection code, so getting rid of it would greatly simplify a lot of it. It would certainly make the logic much easier to follow.

***

For some historical context, #5831 added subtractFeeFromAmount (which was a rebase of #4331) and that points to 2 issues where the proposed use case is sweeping.

achow101 commented:
Closing this in favor of introducing sendall to solve the problem.
